playbook/controllers.go

```diff
@@ -0,0 +1,134 @@
+package playbook
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/duty/types"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+)
+
+type RunResponse struct {
+	RunID    string `json:"run_id"`
+	StartsAt string `json:"starts_at"`
+}
+
+type RunParams struct {
+	ID          uuid.UUID         `json:"id"`
+	ConfigID    uuid.UUID         `json:"config_id"`
+	ComponentID uuid.UUID         `json:"component_id"`
+	Params      map[string]string `json:"params"`
+}
+
+func (r *RunParams) valid() error {
+	if r.ID == uuid.Nil {
+		return fmt.Errorf("playbook id is required")
+	}
+
+	if r.ConfigID == uuid.Nil && r.ComponentID == uuid.Nil {
+		return fmt.Errorf("either config_id or component_id is required")
+	}
+
+	if r.ConfigID != uuid.Nil && r.ComponentID != uuid.Nil {
+		return fmt.Errorf("either config_id or component_id is required")
+	}
+
+	return nil
+}
+
+func (r *RunParams) validateParams(params []v1.PlaybookParameter) error {
+	if len(params) != len(r.Params) {
+		return fmt.Errorf("invalid number of parameters. expected %d, got %d", len(params), len(r.Params))
+	}
+
+	for k := range r.Params {
+		var ok bool
+		for _, p := range params {
+			if k == p.Name {
+				ok = true
+				break
+			}
+		}
+
+		if !ok {
+			return fmt.Errorf("unknown parameter %s", k)
+		}
+	}
+
+	return nil
+}
+
+// HandlePlaybookRun handles playbook run requests.
+func HandlePlaybookRun(c echo.Context) error {
+	ctx := c.(*api.Context)
+
+	var req RunParams
+	if err := c.Bind(&req); err != nil {
+		return c.JSON(http.StatusBadRequest, api.HTTPError{Error: err.Error(), Message: "invalid request"})
+	}
+
+	if err := req.valid(); err != nil {
+		return c.JSON(http.StatusBadRequest, api.HTTPError{Error: err.Error(), Message: "invalid request"})
+	}
+
+	playbook, err := db.FindPlaybook(ctx, req.ID)
```

Comment:
I think the execution of the playbook should be in a separate file/function - especially so that it can be tested.

Reply:
The function that executes is isolated. This controller simply records the run in the db.